src/formats/base_32.cpp

```diff
@@ -0,0 +1,187 @@
+const static size_t data_part_min_length = checksum_length;
+const static size_t base32_min_length = hr_part_min_length + 1 + data_part_min_length;
+const static size_t base32_max_length = 90;
+const static size_t data_part_max_length = base32_max_length - hr_part_min_length - 1;
+
+uint32_t polymod_base32(const data_chunk& values)
+{
+    static const std::vector<uint32_t> gen = 
+        {0x3b6a57b2, 0x26508e6d, 0x1ea119fa, 0x3d4233dd, 0x2a1462b3};
+    uint32_t chk = 1;
+    for (auto v : values)
+    {
+        const uint32_t b = (chk >> 25);
+        chk = (chk & 0x1ffffff) << 5 ^ v;
+        for (auto i = 0; i < 5; ++i)
+        {
+            chk ^= ((b >> i) & 1) ? gen[i] : 0;
+        }
+    }
+    return chk;
+}
+
+data_chunk hrp_expand_base32(const std::string& s)
+{
+    data_chunk result(2 * s.size() + 1);
+    auto iterator = result.begin();
+    for (size_t i = 0; i < s.size(); ++i)
+    {
+        *iterator = s[i] >> 5;
+        ++iterator;
+    }
+    ++iterator; // is already '0'
+    for (size_t i = 0; i < s.size(); ++i)
+    {
+        *iterator = s[i] & 31;
+        ++iterator;
+    }
+    return result;
+}
+
+bool verify_checksum_base32(const std::string& hr_part, const data_chunk& data_part)
+{
+    data_chunk expanded = hrp_expand_base32(hr_part);
+    expanded.insert(expanded.end(), data_part.begin(), data_part.end());
+    return polymod_base32(expanded) == 1;
+}
+
+data_chunk create_checksum_base32(const std::string& hr_part, const data_chunk& data_part)
+{
+    data_chunk values = hrp_expand_base32(hr_part);
+    values.insert(values.begin(), data_part.begin(), data_part.end());
+    values.resize(values.size() + checksum_length);
+    uint32_t polymod = polymod_base32(values) ^ 1;
+    data_chunk result(checksum_length, 0);
+    for (size_t i = 0; i < result.size(); ++i)
+    {
+        result[i] = (polymod >> (5 * (5 - i))) & 31;
+    }
+    return result;
+}
+
+std::string encode_base32(const std::string& hr_part, const data_chunk& data_part)
+{
+    std::string encoded = hr_part + separator;
+    encoded.reserve(encoded.size() + data_part.size() + checksum_length);
+    data_chunk checksum = create_checksum_base32(hr_part, data_part);
+    for (auto value : data_part)
+    {
+        encoded += lut[value];
+    }
+    for (auto value : checksum)
+    {
+        encoded += lut[value];
+    }
+    return encoded;
+}
+
+inline char toLowercase(const char& c)
+{
+    return c + ('a' - 'A');
+}
+
+bool decode_base32(const std::string& in, std::string& hr_part, data_chunk& values)
+{
+    // Basic input check
+    if (in.length() < base32_min_length || in.length() > base32_max_length)
+    {
+        return false;
+    }
+    
+    // Find separator and check valid hr_part and data_part sizes
+    size_t pos_separator = in.rfind(separator);
+    if (pos_separator == std::string::npos 
+        || pos_separator < hr_part_min_length
+        || pos_separator > hr_part_max_length
```

Comment:
The check on this line is redundant because you're already checking for the input string not exceeding the maximum length on line 125 and checking for both parts having the minimum length required.